[REVIEW] Updating DEVELOPER GUIDE to be more precise about the cuml threading model [skip-ci] (#820)
rapidsai/cuml: cpp/DEVELOPER_GUIDE.md

Please start by reading [CONTRIBUTING.md](../CONTRIBUTING.md).

## Thread safety

With the exception of the cumlHandle, cuML algorithms should maintain thread-safety and are, in general, 
assumed to be single threaded. This means they should be able to be called from multiple host threads so 
long as different instances of `cumlHandle` are used.

> Exceptions are made for algorithms that can take advantage of multiple CUDA streams in order to oversubscribe 

Reviewer: One can use multiple CUDA streams from a single CPU thread so better to refer to CPU threads here.

or increase occupancy on a single GPU. In these cases, the use of multiple host threads within cuML algorithms
should be used only to maintain concurrency of the underlying CUDA streams. Multiple host threads should be 
used sparingly, be bounded, and should steer clear of performing CPU-intensive computations.

> A good example of an acceptable use of host threads can be found in [this blog article](https://devblogs.nvidia.com/gpu-pro-tip-cuda-7-streams-simplify-concurrency/)

Reviewer 2: By linking that blog post are you suggesting that cuML should or does use CUDA's per-thread default stream?

Author: @harrism, there were two reasons I linked that blog post, actually. Knowing that we can use the default stream per thread is useful and it might be worth adding it as an option in our build. The other reason I linked to it is that it provides an example of this more general design:

```
cudaStreamSynchronize(streamInCurrentHostThread);

#pragma omp parallel for num_threads(n_threads)
for(int i = 0; i < n; i++) {
    ... possible light cpu pre-processing ...
    cudaStream_t s;
    cudaStreamCreate(&s);
    my_kernel1<<<b, tpb, 0, s>>>(...);
    ...
    ... some possible async d2h / h2d copies ...
    my_kernel2<<<b, tpb, 0, s>>>(...);
    ...
    cudaStreamSynchronize(s);
    ... possible light cpu post-processing ...
}
```

I have not seen an abundance of formal examples for this design. Maybe I should also include this code snippet?

Though cuML's build enables OpenMP by default, cuML algorithms should still function properly even when OpenMP has been

> disabled. For this reason, it would be better to utilize OpenMP rather than raw pthreads for spawning host threads. 

Reviewer 2: This is a little unclear. If you use OpenMP rather than pthreads, will it function properly with OpenMP disabled?

Reviewer: Agree with @harrism this is unclear. I would suggest the following: "To avoid compatibility issues between different threading models the only multi threading programming allowed in cuML is OpenMP."

> The use of threads in thirdparty libraries are okay.

Reviewer 2: ```suggestion
The use of threads in third-party libraries is okay.
```

Reviewer: If this sentence should go in I think it needs to be more specific. E.g. one reason why it is desirable to have OpenMP only as optional dependency is to avoid compatibility issues with a OpenMP runtime a cuML might want to use. E.g. if the application calling into cuML spawns multiple OpenMP threads and let each OpenMP thread work on independent cuML jobs. For that reason third party libraries should avoid to in a dependency for a specific OpenMP runtime.
